### ISSUE TITLE
[java] NonThreadSafeSingleton false positive with double-checked locking

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#4483]
**Affects PMD Version:** 6.x

**Rule:** [NonThreadSafeSingleton](https://docs.pmd-code.org/latest/pmd_rules_java_multithreading.html#nonthreadsafesingleton)

**Description:**

The rule reports a violation even for double checked locking, that is implemented correctly.

**Code Sample demonstrating the issue:**

```java
public class Foo {
    private static volatile Object o;
    public static Object getO() {
        if (o == null) {   // violation
            synchronized(Foo.class) {
                if (o == null) {
                    o = new Object();
                }
            }
        }
        return o;
    }
}
```

**Expected outcome:** No violation

**Note:** This has already been fixed for PMD 7.0.0-rc1 via #3474. This issue is just there for documentation.
